### PR TITLE
[RLlib] Using PG when not doing microbatching kills A2C Performance

### PR DESCRIPTION
**Pull request by avnishn**
The training step behavior when microbatching wasn't enabled
in A2C is to use the PG train step.
But for whatever reason, there is some bug when using the PG
training_step function, compared to when microbatching is
set to the same number of steps as train_batch size

(computing gradients on 1 microbatch)

This doesn't exactly make sense since the doing a
microbatch of 1 is logically the same as doing PG,
but this probably means there is some bug in the PG
training_step function. Its likely gone undetected
since we use PG as a reference implementaion. Its
probably not worth exploring instead of making this
change for the time being, since PG still works well
on simpler environments.

Signed-off-by: avnish <avnish@anyscale.com>

image

here's a tensorboard showing the performance change in the release test that would fail....
blue is before when we would use the pg training_step function. 

orange and red are afterwards where I stopped using the pg training_step function

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gjoliver** [comment]
cool explanation. thanks for catching this. no idea why we fall back to PG.

**Comment by avnishn** [comment]
@gjoliver the new changes that involve A2C and fake gpus are made because before when we would default to the PG training_step function, PG supports fake gpus, but the compute gradients methods of A2C/A3C don't support fake gpus. 

I've added an assertion check in those algorithms to make sure to throw in error if that _fake gpus is set, and modified any relevant tests with regards to that parameter and A2c.

**Comment by avnishn** [comment]
In fact it looks like A3C flat out has no tests whatsoever having _fake_gpus allowed. This probably is allowed and is expected behavior with the A3C policy

**Comment by sven1977** [comment]
Great catch @avnishn ! The reason A3C doesn't have a fake-gpu test is that A3C should probably not be run with GPUs anyways.
The A2C -> PG fallback can be explained with A2C basically being PG plus advantage estimation using a value function (plus training this value function via a vf loss term). Only if microbatching is used, A2C should behave differently from PG.

So just to understand: This PR make A2C always use the microbatch path (micro batch size == 1). Also, we haven't understood why this change is working, but the PG path is not (even though they are believed to be the same)?

**Comment by avnishn** [comment]
That pretty much sums it up.

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Since the ordinary `cartpole-a2c.yaml` is the same - only difference is stopping on iterations/steps - this looks safe.
I think we should do this to have a working test and stop users from running into this behaviour, so we can take our time understanding the bug fully.

**Comment by avnishn** [comment]
+1
